Add new analyzer rules

xtqqczze opened the pull request:
Fix #15498

* Enable [IL3002: Using member with RequiresAssemblyFilesAttribute can break functionality when embedded in a single-file app](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/il3002).
* Enable [CA1418: Use valid platform string](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1418).
* Enable [CA1839: Use 'Environment.ProcessPath'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1839) as _suggestion_.
  * Issue: **#15629**.
* Enable [CA1840: Use 'Environment.CurrentManagedThreadId'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1840).
* Enable [CA1841: Prefer Dictionary.Contains methods](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1841).
* Enable [CA1842: Do not use 'WhenAll' with a single task](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1842).
* Enable [CA1843: Do not use 'WhenAll' with a single task](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1843).
* Enable [CA1844: Provide memory-based overrides of async methods when subclassing 'Stream'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1844).
* Enable [CA1845: Use span-based 'string.Concat'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1845).
* Enable [CA1846: Prefer 'AsSpan' over 'Substring'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1846).
* Enable [CA2250: Use 'ThrowIfCancellationRequested'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2250).
* Enable [CA2251: Use 'string.Equals'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2251).
* Enable [IDE0090: UseNew](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0090) as _suggestion_.
  * Issue: **#15627**.
* Enable [IDE0100: RemoveRedundantEquality](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0100).
* Enable [IDE0110: RemoveUnnecessaryDiscard](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0110) as _suggestion_.
  * Issue: **#15630**.
* Enable [IDE0120: SimplifyLINQExpression](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0120)
* Enable [IDE0130: NamespaceDoesNotMatchFolderStructure](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0130) as _suggestion_.
  * Issue: **#15631**.
* Enable [IDE2000: MultipleBlankLines](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2000)
* Enable [IDE2001: EmbeddedStatementsMustBeOnTheirOwnLine](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2001)
* Enable [IDE2002: ConsecutiveBracesMustNotHaveBlankLinesBetweenThem](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2002)
* Enable [IDE2003: ConsecutiveStatementPlacement](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2003)
* Enable [IDE2004: BlankLineNotAllowedAfterConstructorInitializerColon](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2004)

xtqqczze committed:
Fix case of rule IDs

xtqqczze committed:
Add new rules from dotnet/runtime

https://github.com/dotnet/runtime/commit/3394fd580a8fcf58ac353d0550a22db6f5e50c18
https://github.com/dotnet/runtime/commit/877a8df6716be6a7f98e95d17007619502c43b71
https://github.com/dotnet/runtime/commit/4a2a44a6179576ce7d033cd43626843b39a84ec9
https://github.com/dotnet/runtime/commit/aad916c7111323d4b5744eb2444252d149c57a8a
https://github.com/dotnet/runtime/commit/72ff0df4638ced1cc2b15c92c055f842c51e32d2

iSazonov commented:
/azp list

iSazonov commented:
/azp run PowerShell-CI-windows

iSazonov commented on .globalconfig:
```diff
@@ -181,6 +181,9 @@ dotnet_diagnostic.CA1416.severity = warning
 # CA1417: Do not use 'OutAttribute' on string parameters for P/Invokes
 dotnet_diagnostic.CA1417.severity = warning
 
+# CA1418: Use valid platform string
+dotnet_diagnostic.CA1418.severity = silent
```
I'd set warning on most of new rules if where no complains are raised and open new tracking issue for rest ones so that we review them late.

xtqqczze replied:
Documentation for this rule does not yet exist, I think we should not yet enable it.

https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1418 -> 404

iSazonov replied:
It's temporary. 
It's important for us to make the rules work. If it is silent, then the rule does not work and there is no point in adding it. So I would put warning where possible and for other rules create a tracking issue so that enable them in follow PRs..

xtqqczze committed:
Enable CA1418: Use valid platform string

Enable [CA1418: Use valid platform string](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1418)

xtqqczze committed:
Enable CA1839 as suggestion

Enable [CA1839: Use 'Environment.ProcessPath'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1839)

xtqqczze committed:
Enable IL3002

Enable [IL3002: Using member with RequiresAssemblyFilesAttribute can break functionality when embedded in a single-file app](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/il3002)

xtqqczze committed:
Enable CA1840

Enable [CA1840: Use 'Environment.CurrentManagedThreadId'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1840)

xtqqczze committed:
Enable CA1841: Prefer Dictionary.Contains methods

Enable [CA1841: Prefer Dictionary.Contains methods](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1841)

xtqqczze committed:
Enable CA1842

Enable [CA1842: Do not use 'WhenAll' with a single task](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1842)

xtqqczze committed:
Enable CA1843

Enable [CA1843: Do not use 'WhenAll' with a single task](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1843)

xtqqczze committed:
Enable CA1845: Use span-based 'string.Concat'

Enable [CA1845: Use span-based 'string.Concat'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1845)

xtqqczze committed:
Enable CA1846: Prefer 'AsSpan' over 'Substring'

Enable [CA1846: Prefer 'AsSpan' over 'Substring'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1846)

xtqqczze committed:
Enable CA1844

Enable [CA1844: Provide memory-based overrides of async methods when subclassing 'Stream']((https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1844)

xtqqczze commented on .globalconfig:
```diff
@@ -2,7 +2,7 @@ is_global = true
 
 # CA1000: Do not declare static members on generic types
```
```suggestion
# CA1000: Do not declare static members on generic types
# https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1000
```

@iSazonov I think it would helpful to have documentation links to each rule in `.globalconfig`, do you agree?

rjmholt replied:
I think that would be a good addition

iSazonov replied:
Oh, in the case we should check the links in CI-Static.  There are frequent failures as it is. I don't think we need to make life difficult for us.

xtqqczze replied:
I don't think it should be necessary to check the links in `CI-Static`. The URIs can be provided as a convenience rather than a normative reference.

xtqqczze committed:
Enable CA2250: Use 'ThrowIfCancellationRequested'

Enable [CA2250: Use 'ThrowIfCancellationRequested'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2250)

xtqqczze committed:
Enable CA2251: Use 'string.Equals'

Enable [CA2251: Use 'string.Equals'](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2251)

xtqqczze committed:
Enable IDE0090 as suggestion

Enable [IDE0090: UseNew](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0090)

xtqqczze committed:
Enable IDE0100: RemoveRedundantEquality

Enable [IDE0100: RemoveRedundantEquality](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0100)

xtqqczze committed:
Enable IDE0110 as suggestion

Enable [IDE0110: RemoveUnnecessaryDiscard](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0110)

xtqqczze committed:
Enable IDE0120: SimplifyLINQExpression

Enable [IDE0120: SimplifyLINQExpression](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0120)

xtqqczze committed:
Enable IDE0130 as suggestion

Enable [IDE0130: NamespaceDoesNotMatchFolderStructure](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0130)

xtqqczze committed:
Enable IDE2000: MultipleBlankLines

Enable [IDE2000: MultipleBlankLines](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2000)

xtqqczze committed:
Enable IDE2001

Enable [IDE2001: EmbeddedStatementsMustBeOnTheirOwnLine](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2001)

xtqqczze committed:
Enable IDE2002

Enable [IDE2002: ConsecutiveBracesMustNotHaveBlankLinesBetweenThem](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2002)

xtqqczze committed:
Enable IDE2003: ConsecutiveStatementPlacement

Enable [IDE2003: ConsecutiveStatementPlacement](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2003)

xtqqczze committed:
Enable IDE2004

Enable [IDE2004: BlankLineNotAllowedAfterConstructorInitializerColon](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide2004)

xtqqczze commented:
`PowerShell-CI-macos` failure appears unrelated: `WebListener did not start before the timeout was reached.`

rjmholt commented:
/azp run PowerShell-CI-macos

xtqqczze committed:
Add documentation links

iSazonov commented on .globalconfig:
```diff
@@ -1,1554 +1,2158 @@
 
+# CA1839: Use 'Environment.ProcessPath'
+# https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1839
+dotnet_diagnostic.CA1839.severity = suggestion
```
It seems we fixed this already and can set warning.

iSazonov replied:
Ah, I see a complain for tests.
